File: backend/schemas/transit_fit.py
from pydantic import BaseModel, Field

from schemas.lightcurve import LightCurvePoint


class TransitFitRequest(BaseModel):
    target_id: str
    period: float
    t0: float
    fit_mode: str = "phase_fold"
    bjd_start: float | None = None
    bjd_end: float | None = None
    fit_limb_darkening: bool = False
    fit_window_phase: float = 0.12
    baseline_order: int = 1
    sigma_clip_sigma: float = 4.0
    sigma_clip_iterations: int = 2
    filter_name: str | None = None
    stellar_temperature: float | None = None
    stellar_logg: float | None = None
    stellar_metallicity: float | None = None
    points: list[LightCurvePoint] = Field(default_factory=list)


class TransitFitParameters(BaseModel):
    rp_rs: float
    rp_rs_err: float = 0.0
    a_rs: float
    a_rs_err: float = 0.0
    inclination: float
    inclination_err: float = 0.0
    u1: float
    u1_err: float = 0.0
    u2: float
    u2_err: float = 0.0
    chi_squared: float = 0.0
    reduced_chi_squared: float = 0.0
    degrees_of_freedom: int = 0


class TransitModelCurve(BaseModel):
    phase: list[float]
    flux: list[float]


class TransitFitPreprocessing(BaseModel):
    fit_mode: str = "phase_fold"
    fit_window_phase: float
    bjd_start: float | None = None
    bjd_end: float | None = None
    limb_darkening_source: str | None = None
    limb_darkening_filter: str | None = None
    baseline_order: int
    sigma_clip_sigma: float
    sigma_clip_iterations: int
    retained_points: int = 0
    clipped_points: int = 0


class TransitFitResponse(BaseModel):
    target_id: str
    period: float
    t0: float
    reference_t0: float
    limb_darkening_source: str | None = None
    limb_darkening_filter: str | None = None
    used_batman: bool = True
    used_mcmc: bool = False
    preprocessing: TransitFitPreprocessing
    fitted_params: TransitFitParameters
    initial_params: TransitFitParameters
    model_curve: TransitModelCurve
    initial_curve: TransitModelCurve
    model_time: list[float]
    data_time: list[float]
    data_phase: list[float]
    data_flux: list[float]
    data_error: list[float]
    residuals: list[float]

File: frontend/src/types/transitFit.ts
import type { LightCurvePoint } from './photometry';

export interface TransitFitRequest {
  target_id: string;
  period: number;
  t0: number;
  fit_mode: 'phase_fold' | 'bjd_window';
  bjd_start?: number | null;
  bjd_end?: number | null;
  fit_limb_darkening: boolean;
  fit_window_phase: number;
  baseline_order: number;
  sigma_clip_sigma: number;
  sigma_clip_iterations: number;
  filter_name?: string | null;
  stellar_temperature?: number | null;
  stellar_logg?: number | null;
  stellar_metallicity?: number | null;
  points: LightCurvePoint[];
}

export interface TransitFitParameters {
  rp_rs: number;
  rp_rs_err: number;
  a_rs: number;
  a_rs_err: number;
  inclination: number;
  inclination_err: number;
  u1: number;
  u1_err: number;
  u2: number;
  u2_err: number;
  chi_squared: number;
  reduced_chi_squared: number;
  degrees_of_freedom: number;
}

export interface TransitModelCurve {
  phase: number[];
  flux: number[];
}

export interface TransitFitPreprocessing {
  fit_mode: 'phase_fold' | 'bjd_window';
  fit_window_phase: number;
  bjd_start: number | null;
  bjd_end: number | null;
  limb_darkening_source?: string | null;
  limb_darkening_filter?: string | null;
  baseline_order: number;
  sigma_clip_sigma: number;
  sigma_clip_iterations: number;
  retained_points: number;
  clipped_points: number;
}

export interface TransitFitResponse {
  target_id: string;
  period: number;
  t0: number;
  reference_t0: number;
  limb_darkening_source?: string | null;
  limb_darkening_filter?: string | null;
  used_batman: boolean;
  used_mcmc: boolean;
  preprocessing: TransitFitPreprocessing;
  fitted_params: TransitFitParameters;
  initial_params: TransitFitParameters;
  model_curve: TransitModelCurve;
  initial_curve: TransitModelCurve;
  model_time: number[];
  data_time: number[];
  data_phase: number[];
  data_flux: number[];
  data_error: number[];
  residuals: number[];
}
